xtp/src/libxtp/pmlocalization.cc

```diff
@@ -24,11 +24,379 @@
 #include "votca/xtp/aomatrix.h"
 #include <limits>
 
-#include <votca/tools/eigenio_matrixmarket.h>
-
 namespace votca {
 namespace xtp {
+
 void PMLocalization::computePML(Orbitals &orbitals) {
+
+  if (method_ == "jacobi-sweeps") {
+    XTP_LOG(Log::error, log_)
+        << TimeStamp() << " Using Jacobi-Sweeps" << std::flush;
+    computePML_JS(orbitals);
+  } else if (method_ == "unitary-optimizer") {
+    XTP_LOG(Log::error, log_)
+        << TimeStamp() << " Using Unitary Optimizer" << std::flush;
+    computePML_UT(orbitals);
+  }
+}
+
+double PMLocalization::cost(const Eigen::MatrixXd &W,
+                            const std::vector<Eigen::MatrixXd> &Sat_all,
+                            const Index nat) {
+
+  double Dinv = 0.0;
+  double p = 2.0;  // standard PM
+  for (Index iat = 0; iat < nat; iat++) {
+    Eigen::MatrixXd qw = Sat_all[iat] * W;
+    for (Index i = 0; i < W.cols(); i++) {
+      double Qa = W.col(i).transpose() * qw.col(i);
+      Dinv += std::pow(Qa, p);
+    }
+  }
+  return Dinv;
+}
+
+std::pair<double, Eigen::MatrixXd> PMLocalization::cost_derivative(
+    const Eigen::MatrixXd &W, const std::vector<Eigen::MatrixXd> &Sat_all,
+    const Index nat) {
+  Eigen::MatrixXd Jderiv = Eigen::MatrixXd::Zero(n_occs_, n_occs_);
+  double Dinv = 0.0;
+  double p = 2.0;  // standard PM
+  for (Index iat = 0; iat < nat; iat++) {
+    Eigen::MatrixXd qw = Sat_all[iat] * W;
+    for (Index i = 0; i < W.cols(); i++) {
+      double qwp = W.col(i).transpose() * qw.col(i);
+      Dinv += std::pow(qwp, p);
+      double t = p * std::pow(qwp, p - 1);
+      for (Index j = 0; j < W.cols(); j++) {
+        Jderiv(j, i) += t * qw(j, i);
+      }
+    }
+  }
+  return {Dinv, Jderiv};
+}
+
+Eigen::VectorXd PMLocalization::fit_polynomial(const Eigen::VectorXd &x,
+                                               const Eigen::VectorXd &y) {
+
+  // Fit function to polynomial
+
+  if (x.size() != y.size()) {
+    throw std::runtime_error("x and y have different dimensions!\n");
+  }
+  Index N = x.size();
+  Index deg = N;
+
+  // Form matrix
+  Eigen::MatrixXd A = Eigen::MatrixXd::Zero(N, deg);
+  for (Index i = 0; i < N; i++) {
+    for (Index j = 0; j < deg; j++) {
+      A(i, j) = std::pow(x(i), j);
+    }
+  }
```

Comment:
```suggestion

    for (Index j = 0; j < deg; j++) {
      A.col(i) = x.pow(j);
  }
```

Reply:
Does not work. In general, there I get tons of compiler complaints about `.pow()`. This is not  compute time relevant at all, so I'll leave it as is.